### PR TITLE
Add image loading events to image sources

### PR DESCRIPTION
**Pull request by bartvde**
This follows up on #3248 and adds support for image based sources as well.

### CODE REVIEW

**Comment by ahocevar** [comment]
Looks good to me, please merge.

File: src/ol/renderer/layerrenderer.js [hunk 1]
```diff
@@ -151,7 +151,7 @@ ol.renderer.Layer.prototype.loadImage = function(image) {
     // listener (a noop if the listener was already registered)
     goog.asserts.assert(imageState == ol.ImageState.IDLE ||
         imageState == ol.ImageState.LOADING);
-    goog.events.listenOnce(image, goog.events.EventType.CHANGE,
+    goog.events.listen(image, goog.events.EventType.CHANGE,
```
Comment (elemoine) [inline src/ol/renderer/layerrenderer.js]: Why did you change from `listenOnce` to `listen`?

File: src/ol/source/imagemapguidesource.js [hunk 2]
```diff
@@ -150,6 +153,8 @@ ol.source.ImageMapGuide.prototype.getImage =
     image = new ol.Image(extent, resolution, pixelRatio,
         this.getAttributions(), imageUrl, this.crossOrigin_,
         this.imageLoadFunction_);
+    goog.events.listen(image, goog.events.EventType.CHANGE,
+        this.handleImageChange, false, this);
```
Comment (elemoine) [inline src/ol/source/imagemapguidesource.js]: When is the listener removed?

File: src/ol/renderer/layerrenderer.js (hunk 1 above)
Comment (bartvde) [inline src/ol/renderer/layerrenderer.js]: Otherwise the change to ```ol.ImageState.LOADING``` is not listened for
See the corresponding similar change here: https://github.com/tschaub/ol3/commit/bf35b40b11790cf3e6dcd693bd663a0cdba4016b

File: src/ol/source/imagemapguidesource.js (hunk 2 above)
Comment (bartvde) [inline src/ol/source/imagemapguidesource.js]: in ```disposeInternal``` of ```ol.Image``` similar approach used as here: https://github.com/openlayers/ol3/pull/3248

File: src/ol/renderer/layerrenderer.js (hunk 1 above)
Comment (bartvde) [inline src/ol/renderer/layerrenderer.js]: What I meant to say is otherwise the image won't be rendered because of extra call to changed()  needed for the load events

Comment (elemoine) [inline src/ol/renderer/layerrenderer.js]: Where do we call `dispose` on the image? (Sorry for being insistant here but I want to make sure we're not introducing leaks.)

Comment (bartvde) [inline src/ol/renderer/layerrenderer.js]: jn the destructor of the image

Comment (elemoine) [inline src/ol/renderer/layerrenderer.js]: What destructor? I don't think we currently call `goog.dispose` on `ol.Image` objects, making `ol.Image#disposeInternal` useless.

Comment (elemoine) [inline src/ol/renderer/layerrenderer.js]: Note that in the cases of tiles (`ol.Tile`) we at least call `dispose` on expired tiles. See https://github.com/openlayers/ol3/blob/master/src/ol/tilecache.jshttps://github.com/openlayers/ol3/blob/master/src/ol/tilecache.js.

Comment (bartvde) [inline src/ol/renderer/layerrenderer.js]: Ah good point, so yes, this is an issue indeed

Comment (elemoine) [inline src/ol/renderer/layerrenderer.js]: > Ah good point, so yes, this is an issue indeed

I am not sure :-)

So, with your changes, "change" listeners are not removed from `ol.Image` objects. But this may not be a problem, as it may not prevent the GC from garbage-collecting `ol.Image` objects that aren't used anymore. I'd suggest verifying this before merging this.

Comment (bartvde) [inline src/ol/renderer/layerrenderer.js]: Would calling registerDisposable on the image objects make sense in the sources or not?

Comment (elemoine) [inline src/ol/renderer/layerrenderer.js]: > Would calling registerDisposable on the image objects make sense in the sources or not?

The problem is that you don't know when you may dispose of the image. Maybe the renderer still uses that image for example…

Comment (bartvde) [inline src/ol/renderer/layerrenderer.js]: But registerDisposable only associates the image with the source so that the image gets disposed when the source gets disposed? Though I am not sure when or if the source will be disposed.
I'll investigate a bit more to confirm the GC cleans up.

Comment (elemoine) [inline src/ol/renderer/layerrenderer.js]: > But registerDisposable only associates the image with the source so that the image gets disposed when the source gets disposed?

True.



> Though I am not sure when or if the source will be disposed.

It won't be :-)

Comment (bartvde) [inline src/ol/renderer/layerrenderer.js]: @elemoine I've debugged in the layer renderer and after a zoom action there is only 1 change listener (in the line before the change listener from the renderer gets added), i.e. the one from the image wms source.

Is that enough confirmation that the listener gets garbage collected or were you thinking something else to verify this?

Comment (elemoine) [inline src/ol/renderer/layerrenderer.js]: What I wanted to verify is that there is no memory leaks as new images are requested. The browser should free the memory associated to old images.

Comment (bartvde) [inline src/ol/renderer/layerrenderer.js]: Okay I've verified this and can't see any difference in memory consumption before and after my change.

Comment (elemoine) [inline src/ol/renderer/layerrenderer.js]: Thanks for checking @bartvde!

Comment (elemoine) [inline src/ol/renderer/layerrenderer.js]: And thanks for removing the useless `disposeInternal` function.

Comment (bartvde) [inline src/ol/renderer/layerrenderer.js]: thanks for pointing out that it was useless ;-)
